pkg/ingester/ingester_v2.go

```
@@ -1141,3 +1177,75 @@ sendLoop:
 	// wait for ongoing workers to finish.
 	wg.Wait()
 }
+
+// This method is called as part of Lifecycler's shutdown, to flush all data.
+// Lifecycler shutdown happens as part of Ingester shutdown (see stoppingV2 method).
+// Samples are not received at this stage. Shipping service has already been stopped as well.
+func (i *Ingester) v2LifecyclerFlush() {
+	level.Info(util.Logger).Log("msg", "starting to flush and ship TSDB blocks")
+
+	ctx := context.Background()
+
+	i.compactBlocks(ctx, true)
+	if i.cfg.TSDBConfig.ShipInterval > 0 {
+		i.shipBlocks(ctx)
+	}
+
+	level.Info(util.Logger).Log("msg", "finished flushing and shipping TSDB blocks")
+}
+
+// Blocks version of Flush handler. It force-compacts blocks, and triggers shipping.
+func (i *Ingester) v2FlushHandler(w http.ResponseWriter, _ *http.Request) {
+	go func() {
+		ingCtx := i.BasicService.ServiceContext()
+		if ingCtx == nil || ingCtx.Err() != nil {
+			// Not started or already finished.
```

We should log an error here, otherwise we silently discard the flush request.